envoy/config/filter/http/jwt_authn/v2/config.proto

```diff
@@ -0,0 +1,197 @@
+
+syntax = "proto3";
+
+package envoy.config.filter.http.jwt_authn.v2;
+
+import "envoy/api/v2/route/route.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/wrappers.proto";
+import "validate/validate.proto";
+
+// JSON Web Token (JWT) token format for authentication is defined
+// `here <https://tools.ietf.org/html/rfc7519>`_.
+// See `OAuth2.0 <https://tools.ietf.org/html/rfc6749>`_ and
+// `OIDC1.0 <http://openid.net/connect>`_ for the authentication flow.
+//
+// Example,::
+//
+//     issuer: https://example.com
+//     audiences:
+//     - bookstore_android.apps.googleusercontent.com
+//       bookstore_web.apps.googleusercontent.com
+//     jwks_uri: https://example.com/.well-known/jwks.json
+//
+// [#not-implemented-hide:]
+message JSONWebToken {
+  // Identifies the principal that issued the JWT. See
+  // `here <https://tools.ietf.org/html/rfc7519#section-4.1.1>`_.
+  // Usually a URL or an email address.
+  //
+  // Example: https://securetoken.google.com
+  // Example: 1234567-compute@developer.gserviceaccount.com
+  //
+  string issuer = 1;
+
+  // The list of JWT `audiences
+  // <https://tools.ietf.org/html/rfc7519#section-4.1.3>`_. that are allowed to
+  // access. A JWT containing any of these audiences will be accepted.
+  //
+  // Example::
+  //
+  //     audiences:
+  //     - bookstore_android.apps.googleusercontent.com
+  //       bookstore_web.apps.googleusercontent.com
+  //
+  repeated string audiences = 2;
+
+  // URL of the provider's public key set to validate signature of the
+  // JWT. See `OpenID
+  // Discovery
+  // <https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata>`_
+  //
+  // Example: https://www.googleapis.com/oauth2/v1/certs
+  string jwks_uri = 3;
+
+  // This field is specific for Envoy proxy implementation.
+  // It is the cluster name in the Envoy "cluster_manager" config section.
+  // In order for Envoy to call "jwks_uri", its host has to be specified
+  // as a "cluster" in the config for each jwks_uri.
+  string jwks_uri_cluster = 4;
```

Comment:
This seems strange. Yes, Envoy needs a cluster to call out to, but in that case, `jwks_uri` should just be the path suffix, the host part of the URL isn't used?

Also, how is this supposed to be used normally? Are we going to be calling out to a dedicated internal server for jwks_uri, or is this arbitrary Internet callout?

Reply:
Arbitrary Internet callout

I like to keep pks_uri so that it not only specify the path,  also specify protocol, such as http or https

cluster only specify port:  unusually is tcp://x.x.x.x:443

Reply:
While cluster does know whether it is a SSL or not (`tls_context`), we still need protocol and hostname from jwks_uri for `:scheme` and `:authority` header.

Reply:
One should use gwks_uri with discovery service to find the cluster. I don't think the JWT itself should have this information. URI can contain arbitrary amount of information, such as base?cluster=xxx

Reply:
To me, this is a pretty weird situation, where we might not want to specify a fixed cluster if it's arbitrary Internet callout. It's similar to what's done for the gRPC service https://github.com/envoyproxy/data-plane-api/blob/master/envoy/api/v2/core/grpc_service.proto#L32, where we have a cluster-less URI and the library just connects directly. We probably would need to make use of the pending cluster late binding support in https://github.com/envoyproxy/envoy/pull/2740, but we could create a dynamic cluster on-the-fly to use for this URI. @mattklein123 also for thoughts.

Reply:
We will implemented this in phases.  In phase 1, before Envoy supports Lazy config loading or dynamic cluster,  a dedicated jwks cluster is required.  It can be appended in the jwks_uri as "JWKS_URI?cluster=name".  So this field is removed.
In the phase2,  if lazy config loading is supported, this requirement is removed.

Reply:
I strongly disagree with the idea to include `?cluster=` in to `jwks_uri`, `jwks_uri` can be a arbitrary URI (including query parameter), the cluster is really just envoy term, so it has to be a separate field and has nothing to do with `jwks_uri`.

Also in the case that control-plane handles OIDC discovery, `jwks_uri` is coming directly from the discovery document, while cluster is generated by control-plane, please do not mix them.

@htuch this is just like a `EnvoyGrpc` message, (i.e. cluster name). if we want a general EnvoyHttp alike message in core that is a different issue.

Reply:
I think we should have an `EnvoyHttp` source or the like. This is a weird hack in the API IMHO, we shouldn't do it as is.

Reply:
This is somewhat related to https://github.com/envoyproxy/envoy/issues/1606. I agree with @htuch that we should add an `EnvoyHttp` message to allow for generic callout. For now that message can just have a static cluster option, but we can enhance it later to do completely independent DNS resolution/caching and generic calls.

Reply:
@qiwzhang Can you add an `EnvoyHttp` message into envoy core?